### ISSUE TITLE
Way to use precise and generic overloads together?

### ISSUE THREAD

**Issue by Dreamsorcerer** [python/mypy#11005]
I've just come across this function:
```
def str_if_bytes(value):
    return value.decode("utf-8", errors="replace") if isinstance(value, bytes) else value
```

And trying to type it, I have attempted this:
```python
@overload
def str_if_bytes(value: bytes) -> str:
    ...
@overload
def str_if_bytes(value: _T) -> _T:
    ...
def str_if_bytes(value: object) -> object:
    return value.decode("utf-8", errors="replace") if isinstance(value, bytes) else value
```

But, mypy gives this error:
`Overloaded function signatures 1 and 2 overlap with incompatible return types  [misc]`

Is it not possible to specify more precise overloads before more generic ones?

**Comment by erictraut** [comment]
This looks like a bug in mypy's logic for detecting overlapping overload signatures. Pyright doesn't emit an error in this case. You might need to use a `# type: ignore` to work around it for now.

**Comment by hauntsaninja** [comment]
Yeah, `Overloaded function signatures 1 and 2 overlap with incompatible return types` is something that one sometimes just needs to type ignore; mypy should do the correct thing once you've done so

**Comment by AlexWaygood** [comment]
This question has been answered, so I'm closing the issue. I'm confident that there are other issues that discuss the correctness of mypy's overlapping-overloads logic in more depth.
